lib/Network/OAuth2/Server/API.hs

```diff
@@ -322,50 +322,80 @@ handleShib
 handleShib f (Just u) (Just s) = f u s
 handleShib _ _        _        = error "Expected Shibbloleth headers"
 
--- | Receive requests from clients and display the authorization user interface
+-- | Implement the OAuth2 authorize endpoint.
 --
--- TODO: Handle the validation of things more nicely here, preferably shifting
--- them out of here entirely.
+--   This handler must be protected by Shibboleth (or other mechanism in the
+--   front-end proxy). It decodes the client request and presents a UI allowing
+--   the user to approve or reject a grant request.
+--
+--   TODO: Handle the validation of things more nicely here, preferably
+--   shifting them out of here entirely.
+--
+--   http://tools.ietf.org/html/rfc6749#section-3.1
 authorizeEndpoint
     :: ( MonadIO m
        , MonadBaseControl IO m
        , MonadError ServantErr m
        )
     => Pool Connection
-    -> UserID
-    -> Scope
-    -> Maybe ResponseType
-    -> Maybe ClientID
-    -> Maybe RedirectURI
-    -> Maybe Scope
-    -> Maybe ClientState
+    -> UserID -- ^ Authenticated user
+    -> Scope  -- ^ Authenticated permissions
+    -> Maybe ResponseType -- ^ Requested response type.
+    -> Maybe ClientID     -- ^ Requesting Client ID.
+    -> Maybe RedirectURI  -- ^ Requested redirect URI.
+    -> Maybe Scope        -- ^ Requested scope.
+    -> Maybe ClientState  -- ^ State from requesting client.
     -> m Html
-authorizeEndpoint pool user_id permissions rt c_id' redirect sc' st = do
-    case rt of
-        Nothing -> error "Response type code is missing"
-        Just ResponseTypeCode -> return ()
-        Just x -> error $ "Invalid response type: " <> show x
-    sc <- case sc' of
-        Nothing -> error "Scope is missing"
-        Just sc -> if sc `compatibleScope` permissions then return sc else error "NOOOOO"
-    c_id <- case c_id' of
-        Nothing -> error "ClientID is missing"
-        Just c_id -> return c_id
-    res <- liftIO $ storeLookupClient pool c_id
-    ClientDetails{..} <- case res of
-        Nothing -> error $ "no client found with id" <> show c_id
+authorizeEndpoint pool user_id user_perms resp_type' cid' redirect scope' state' = do
+    -- Required: a supported ResponseTypeCode value.
+    response_type <- case resp_type' of
+        Just ResponseTypeCode -> return ResponseTypeCode
+        Just _ -> throwOAuth2Error $ OAuth2Error UnsupportedResponseType
+                                     (preview errorDescription "The authorization grant type is not supported by the authorization server.")
+                                     Nothing
+        Nothing -> throwOAuth2Error $ OAuth2Error InvalidRequest
+                                      (preview errorDescription "The request is missing the response_type parameter.")
```

Comment:
Is this necessarily true? If there is an invalid response_type methinks it will hit this branch?